### PR TITLE
Add algorithm to compute the generating function of plane partitions

### PR DESCRIPTION
**Pull request by EleriasQueflunn**
This PR implements an efficient algorithm to compute the generating polynomial $N_q(a, b, c)$ for plane partitions contained in an $a \times b \times c$ box, based on MacMahon's formula:

$$N_q(a, b, c) = \prod_{i=1}^{a} \prod_{j=1}^{b} \frac{1-q^{i+j+c-1}}{1-q^{i+j-1}}$$

### Mathematical approach
To avoid the computational cost of naive polynomial multiplication and division, we factorize the generating function into cyclotomic polynomials $\Phi_d(q)$:

$$N_q(a, b, c) = \prod_{d=1}^{a+b+c-1} \Phi_d(q)^{e_d}$$

The exponents $e_d$ can be computed directly : 

$$e_d = \sum_{i=0}^{a-1} \left( \left\lfloor \frac{b+c+i}{d} \right\rfloor - \left\lfloor \frac{c+i}{d} \right\rfloor + \left\lfloor \frac{i}{d} \right\rfloor - \left\lfloor \frac{b+i}{d} \right\rfloor \right)$$

#### Proof of this formula
Using the identity $1-q^n = \prod_{d|n} \Phi_d(q)$, the expression can be rewritten as:

$$N_q(a, b, c) = \prod_{i=1}^a \prod_{j=1}^b \frac{\prod_{d|i+j+c-1} \Phi_d(q)}{\prod_{d|i+j-1} \Phi_d(q)}$$

For a fixed $i$ and $d$, the factor $\Phi_d(q)$ appears in the numerator for each $j$ such that $d$ divides $i+j+c-1$. This corresponds to the number of multiples of $d$ in the interval $[i+c, i+c+b-1]$, which is:
$$\left\lfloor \frac{i+c+b-1}{d} \right\rfloor - \left\lfloor \frac{i+c-1}{d} \right\rfloor$$

Similarly, the number of times $\Phi_d(q)$ appears in the denominator is:
$$\left\lfloor \frac{i+b-1}{d} \right\rfloor - \left\lfloor \frac{i-1}{d} \right\rfloor$$

Summing the difference over $i=1$ to $a$ gives the total exponent $e_d$:
$$e_d = \sum_{i=1}^a \left( \left\lfloor \frac{i+c+b-1}{d} \right\rfloor - \left\lfloor \frac{i+c-1}{d} \right\rfloor - \left\lfloor \frac{i+b-1}{d} \right\rfloor + \left\lfloor \frac{i-1}{d} \right\rfloor \right)$$

We get the formula by the change of index $i \leftarrow i-1$.

### Implementation details
This approach is analogous to the existing implementation of the $q$-binomial coefficient in SageMath.

Tests and edge cases (empty boxes, 1D boxes checking against `q_binomial`, and evaluation at $q=1$ for cardinality) have been added to ensure correctness.

### CODE REVIEW

**Comment by mantepse** [comment]
Looks good in principle!  I think it may be slightly better to call it `generating_series`:
There is currently no `generating_polynomial` in sage, but there are a few methods `generating_series` in the same spirit:
    * `SetsWithGrading.ParentMethods`,
    * `PieriFactors`,
    * `SubMultiset_s`, unfortunately, with a typo (`generating_serie`, and only for this parent).

The findstat interface uses `generating_functions`, which is my fault and unfortunate.  I never thought of a good design for this functionality.

It may make sense to put `PlanePartitions()` - if there are no arguments - into the category `SetsWithGrading`.  I think that in this case the grading by size is most natural.

Apart from that, it is a bit odd that the method accepts a variable, but really wants a univariate polynomial ring.  Could you add a fallback if the parent of the variable is not a univariate polynomial ring, i.e., `isinstance(R, PolynomialRing_generic)`?

**Comment by EleriasQueflunn** [comment]
Thank you for your review. I updated the name of the function and I switched to using the function cyclotomic_value (as in `q_binomial`). Now the code works for any q.

**Review by mantepse — APPROVED** [review]
Thank you!

It seems that `generating_series(1)` is now actually (significantly) faster than `cardinality()` in non-trivial cases.  For example:
```
sage: p = PlanePartitions([20,30,50])
sage: %time p.generating_series(1)
CPU times: user 8.42 ms, sys: 0 ns, total: 8.42 ms
Wall time: 8.4 ms
236989731399480431259239914608914674522826332231416031173240008907323814107135176574787528986746496699243837350800044940847618341156776044639493157998725247438824092889792580359017601860537381573768158709802800733057658368767751420999824261157320663064691042880615547837848453846562260134908438446820070522599243776
sage: %time p.cardinality()
CPU times: user 41.8 ms, sys: 889 μs, total: 42.7 ms
Wall time: 42.6 ms
236989731399480431259239914608914674522826332231416031173240008907323814107135176574787528986746496699243837350800044940847618341156776044639493157998725247438824092889792580359017601860537381573768158709802800733057658368767751420999824261157320663064691042880615547837848453846562260134908438446820070522599243776
```

**Comment by EleriasQueflunn** [comment]
Oh, that's unexpected. The function cardinality() should be faster if we use the formula $$N(a, b, c) = \prod_{i=1}^a \prod_{j=1}^b \frac{c+i+j-1}{i+j-1}$$ with $a \leq b \leq c$ instead of $$N(a, b, c) = \prod_{i=1}^a \prod_{j=1}^b \prod_{k=1}^c \frac{i+j+k-1}{i+j+k-2}$$ which is used currently. Would it be okay to include this optimization in the current pull request, or should I open a separate ticket for it?

**Comment by mantepse** [comment]
As you prefer!

**Comment by EleriasQueflunn** [comment]
That's done. Now ```cardinality``` is faster than ```generating_series```. On the same example : 
```Python
sage: p = PlanePartitions([20,30,50])
sage: %time p.generating_series(1)
CPU times: user 3.86 ms, sys: 0 ns, total: 3.86 ms
Wall time: 3.86 ms
236989731399480431259239914608914674522826332231416031173240008907323814107135176574787528986746496699243837350800044940847618341156776044639493157998725247438824092889792580359017601860537381573768158709802800733057658368767751420999824261157320663064691042880615547837848453846562260134908438446820070522599243776
sage: %time p.cardinality()
CPU times: user 569 μs, sys: 0 ns, total: 569 μs
Wall time: 572 μs
236989731399480431259239914608914674522826332231416031173240008907323814107135176574787528986746496699243837350800044940847618341156776044639493157998725247438824092889792580359017601860537381573768158709802800733057658368767751420999824261157320663064691042880615547837848453846562260134908438446820070522599243776
```

File: src/sage/combinat/plane_partition.py [hunk 1]
```diff
@@ -1679,26 +1679,102 @@ def cardinality(self) -> Integer:
...
+            1
+
+            sage: PlanePartitions([1, 8, 5]).generating_series() == q_binomial(8+5, 5)
+            True
+
+            sage: P = PlanePartitions([4, 6, 3])
+            sage: P.cardinality() == P.generating_series()(1)
+            True
+        """
+
```
Comment (mantepse) [inline src/sage/combinat/plane_partition.py]:
```suggestion
```

**Comment by mantepse** [comment]
Thank you!  Sorry, I overlooked a blank line that shouldn't be there.

**Review by mantepse — APPROVED** [review]
thank you
